### PR TITLE
Hopefully fixed AI fleet strength estimation for new upgrade system

### PR DESCRIPTION
**Pull request by Morlic-fo**

### CODE REVIEW

File: default/AI/AIstate.py [hunk 1]
```diff
@@ -662,6 +661,14 @@ def rate_fleet(self, fleet_id, enemy_stats=None, ship_info=None):
                 shields = ship.currentMeterValue(fo.meterType.shield)
                 stats['structure'] = structure
                 stats['shields'] = shields
+                stats['attack'] = 0
+                stats['attacks'] = {}
+                design = fo.getShipDesign(design_id)
+                for partname in design.parts:
```
Comment (Vezzra) [inline default/AI/AIstate.py:670]: You need to check if `fo.getShipDesign(design_id)` returned a valid design. I'm getting these errors when testing:

```
2015-12-26 21:10:03.836602 [error] AI : PythonLoggingWrapper.cpp:41 : Traceback (most recent call last):
2015-12-26 21:10:03.836644 [error] AI : PythonLoggingWrapper.cpp:41 :   File "/Users/user/SoftwareProjekte/FO/src/Xcode/build/Test/FreeOrion.app/Contents/Resources/default/AI/freeorion_tools.py", line 92, in wrapper
2015-12-26 21:10:03.836723 [error] AI : PythonLoggingWrapper.cpp:41 :     return function(*args, **kw)
2015-12-26 21:10:03.836758 [error] AI : PythonLoggingWrapper.cpp:41 :   File "/Users/user/SoftwareProjekte/FO/src/Xcode/build/Test/FreeOrion.app/Contents/Resources/default/AI/freeorion_debug/listeners.py", line 42, in wrapper
2015-12-26 21:10:03.836800 [error] AI : PythonLoggingWrapper.cpp:41 :     res = funct(*args)
2015-12-26 21:10:03.836827 [error] AI : PythonLoggingWrapper.cpp:41 :   File "/Users/user/SoftwareProjekte/FO/src/Xcode/build/Test/FreeOrion.app/Contents/Resources/default/AI/FreeOrionAI.py", line 195, in generateOrders
2015-12-26 21:10:03.836920 [error] AI : PythonLoggingWrapper.cpp:41 :     foAIstate.refresh()  # checks exploration border & clears roles/missions of missing fleets & updates fleet locs & threats
2015-12-26 21:10:03.836949 [error] AI : PythonLoggingWrapper.cpp:41 :   File "/Users/user/SoftwareProjekte/FO/src/Xcode/build/Test/FreeOrion.app/Contents/Resources/default/AI/AIstate.py", line 200, in refresh
2015-12-26 21:10:03.837041 [error] AI : PythonLoggingWrapper.cpp:41 :     self.update_system_status()
2015-12-26 21:10:03.837076 [error] AI : PythonLoggingWrapper.cpp:41 :   File "/Users/user/SoftwareProjekte/FO/src/Xcode/build/Test/FreeOrion.app/Contents/Resources/default/AI/AIstate.py", line 326, in update_system_status
2015-12-26 21:10:03.837211 [error] AI : PythonLoggingWrapper.cpp:41 :     e_rating = self.rate_fleet(fleet_id)
2015-12-26 21:10:03.837253 [error] AI : PythonLoggingWrapper.cpp:41 :   File "/Users/user/SoftwareProjekte/FO/src/Xcode/build/Test/FreeOrion.app/Contents/Resources/default/AI/AIstate.py", line 667, in rate_fleet
2015-12-26 21:10:03.837504 [error] AI : PythonLoggingWrapper.cpp:41 :     for partname in design.parts:
2015-12-26 21:10:03.837541 [error] AI : PythonLoggingWrapper.cpp:41 : AttributeError: 'NoneType' object has no attribute 'parts'
```

I guess the case shouldn't actually occur, but it does, so you need to catch it. Of course, tracking down why `fo.getShipDesign(design_id)` returns `None` in certain cases probably needs to be done at some point...

Comment (Morlic-fo) [inline default/AI/AIstate.py:670]: Did the bug occur in multiplayer? I tend to forget that there exists this weird bug with the design ids which seems impossible to track down and so far never seemed to occur in single player...
Or did this actually happen in singleplayer?

Will fix...

Comment (Morlic-fo) [inline default/AI/AIstate.py:670]: > Did the bug occur in multiplayer?

Actually, even more specific: I would assume the bug happened with you as moderator?

**Comment by geoffthemedio** [comment]
If a moderator clicks turn before that AIs are done their turn, stuff will likely break...

Using the auto-cycle turns option should avoid that happening.

**Comment by Morlic-fo** [comment]
> Using the auto-cycle turns option should avoid that happening.

Not the case. I never manually advance the turns but use the auto-cycling option. The bug still occurs.

Manually advancing the turns before AI orders definitely increases the chance of it happening by some magnitudes though. So if we assume this to be the underlying issue, perhaps there is some bug with the auto-cycling which allows it to prematurely advance under certain conditions?

File: default/AI/AIstate.py [hunk 2]
```diff
@@ -662,6 +661,18 @@ def rate_fleet(self, fleet_id, enemy_stats=None, ship_info=None):
...
+                design = ship.design
+                if not design:
+                    print "ERROR: Ship has no valid design :)"
+                    print "If this happened in a moderated game: Screw those. The AI simply does not work with them."
+                    continue
+                for partname in design.parts:
+                    if partname and fo.getPartType(partname).partClass == fo.shipPartClass.shortRange:
+                        damage = ship.currentPartMeterValue(fo.meterType.capacity, partname)
+                        stats['attack'] += damage
+                        stats['attacks'][damage] = stats['attacks'].get(damage, 0) + 1
```
Comment (Cjkjvfnby) [inline default/AI/AIstate.py:674]: You add two new fields to stats.  Does some one use them?

Comment (Morlic-fo) [inline default/AI/AIstate.py:674]: No, I hoped adding random entries to a dict would make all the difference. Is this not how programming works?

The fields are already defined at `stats = dict(self.get_weighted_design_stats(design_id, species_name))` and are used in the following lines (e.g. look at 2 lines below this comment.

I merely updated them to the correct values we want to use as design stats simply are no longer reliable due to tech upgrades and we need to query the ship meters instead.

Comment (Cjkjvfnby) [inline default/AI/AIstate.py:674]: Do we need lines 664, 665 if they are already defined? No need yo define it twice.

Comment (Morlic-fo) [inline default/AI/AIstate.py:674]: I do not redefine it, I reset its values as after `get_weighted_design_stats(design_id, species_name)` they contain the corresponding stats of the design.

Probably, we could ignore the design completely and only loop over ship stats but that is cleanup beyond the scope of this fix.
Of course, at some point in time, we also want to adress that part of the code but add it onto the list of 100 parts of the AI code which need to be cleaned up, reformatted, documented and so on...

I am busy rewriting other parts of the code atm, so I won't bother with cleaning up this part. Writing 2 clean lines in an environment of 1000 messy ones is a waste of time if it needs to think about longer than 10 seconds... Better do it all at once.
As long as the bugfix works, that should be all required to merge this PR. If you want to clean the mess up, feel free to do so, but probably should be done in a later PR as this is a very relevant bug.

**Comment by Morlic-fo** [comment]
Also, I probably want to squash commits into one before merging. For sake of the comments, I will wait for the final "I want to merge this" call from Vezzra before doing so.

File: default/AI/AIstate.py (hunk 1 above)
Comment (Vezzra) [inline default/AI/AIstate.py:670]: Nope, the bug occurred in a default quickstart game. I switched on auto-advance turns, built a super-tester takeover and just let the game run, for ~60 turns. By that time I got frequent chat error messages from AIs 1 and 3, checked the logs and there they were, in abundance.

So whatever is going on here has nothing to do with multiplayer/moderator games. I've also found a lot of error messages that a certain ship design could not be found, which occurred somewhere during effect processing. Looks as if some ships fly around which don't have a valid design assigned...

**Comment by Vezzra** [comment]
Ok, run a test with the updated PR. Default single player qickstart game, built a super-tester takeover and let the game run for ~100 turns. Got a lot of these two errors:

```
2015-12-27 18:49:38.630178 [error] AI : Universe.cpp:2092 : GetEffectsAndTargets couldn't get ShipDesign
```

and

```
2015-12-27 18:49:38.697487 [debug] AI : ERROR: Ship has no valid design :)
2015-12-27 18:49:38.697530 [debug] AI : If this happened in a moderated game: Screw those. The AI simply does not work with them.
```

The second one indicating that the fix works, but the underlying issue of course remains. The first one is what I referred to [in this comment](https://github.com/freeorion/freeorion/pull/429#discussion_r48456787). However, it's obviously not happening somewhere during effect processing, but apparently when the AI calls some interface function that invokes some of the effect processing code with a non-existing ship design.

I've zipped up the log files, can be [downloaded here](https://dl.dropboxusercontent.com/u/50689031/logfiles.zip).

**Comment by Vezzra** [comment]
> Also, I probably want to squash commits into one before merging.

Ah, good you mentioned that, otherwise I'd have merged the PR now :smiley:

> For sake of the comments, I will wait for the final "I want to merge this" call from Vezzra before doing so.

Well, the fix works, at least on OSX. So I think we can merge this, as the issue with the AI trying to access non-existing ship designs isn't caused by this PR. @Morlic-fo, go ahead with cleaning up the PR and give us a heads up when you're done.

**Comment by Morlic-fo** [comment]
Cleaned up.

**Comment by Vezzra** [comment]
[Posted a follow up](http://freeorion.org/forum/viewtopic.php?f=9&t=9877) on the issue with the ships without design on the forum.

File: default/AI/AIstate.py (hunk 2 above)
Comment (Cjkjvfnby) [inline default/AI/AIstate.py:674]: :(

**Pull request merged**